homeassistant/components/xiaomi_miio/sensor.py

```diff
@@ -145,21 +195,129 @@ async def async_setup_entry(hass, config_entry, async_add_entities):
                         for variable in sensor_variables
                     ]
                 )
-
-    if config_entry.data[CONF_FLOW_TYPE] == CONF_DEVICE:
+    elif config_entry.data[CONF_FLOW_TYPE] == CONF_DEVICE:
         host = config_entry.data[CONF_HOST]
         token = config_entry.data[CONF_TOKEN]
-        name = config_entry.title
-        unique_id = config_entry.unique_id
-
-        _LOGGER.debug("Initializing with host %s (token %s...)", host, token[:5])
+        if model in MODELS_HUMIDIFIER_MIOT:
+            device = hass.data[DOMAIN][config_entry.entry_id][KEY_DEVICE]
+            coordinator = hass.data[DOMAIN][config_entry.entry_id][KEY_COORDINATOR]
+            attributes = HUMIDIFIER_ATTRIBUTES_MIOT
+        elif model.startswith("zhimi.humidifier."):
+            device = hass.data[DOMAIN][config_entry.entry_id][KEY_DEVICE]
+            coordinator = hass.data[DOMAIN][config_entry.entry_id][KEY_COORDINATOR]
+            attributes = HUMIDIFIER_ATTRIBUTES
+        if device:
+            for attribute in attributes:
+                entities.append(
+                    XiaomiHumidifierSensor(
+                        f"{config_entry.title}_{attribute}",
+                        device,
+                        config_entry,
+                        f"{attribute}_{config_entry.unique_id}",
+                        attribute,
+                    )
+                )
+        else:
+            unique_id = config_entry.unique_id
+            name = config_entry.title
+            _LOGGER.debug("Initializing with host %s (token %s...)", host, token[:5])
 
-        device = AirQualityMonitor(host, token)
-        entities.append(XiaomiAirQualityMonitor(name, device, config_entry, unique_id))
+            device = AirQualityMonitor(host, token)
+            entities.append(
+                XiaomiAirQualityMonitor(name, device, config_entry, unique_id)
+            )
 
     async_add_entities(entities, update_before_add=True)
 
 
+class XiaomiHumidifierSensor(XiaomiMiioEntity, SensorEntity):
```

Comment:
Is there something so special for humidifer sensors versus regular sensors, that this should not be a `XiaomiSensor` and re-used by other platforms?